airflow/jobs/scheduler_job_runner.py

```diff
@@ -1844,6 +1895,58 @@ def _executor_to_tis(self, tis: list[TaskInstance]) -> dict[BaseExecutor, list[T
                 executor = str(ti.executor)
             else:
                 executor = None
-            _executor_to_tis[ExecutorLoader.load_executor(executor)].append(ti)
+
+            if executor_obj := self._try_to_load_executor(executor):
+                _executor_to_tis[executor_obj].append(ti)
+            else:
+                continue
 
         return _executor_to_tis
+
+    def _slots_free_for_tis(self, tis: list[TaskInstance]) -> set[TaskInstance]:
+        """Return TIs that we have slots available for."""
+        # First get a mapping of executor names to slots they have available
+        executor_to_slots_available: dict[ExecutorName, int] = {}
+        for executor in self.job.executors:
+            # All executors should have a name if they are initted from the executor_loader. But we need to
+            # check for None to make mypy happy.
```

Comment:
more self-documenting way to do this is 
```python
if TYPE_CHECKING:
    assert executor.name
```

Reply:
Neat idea!  I've never used TYPE_CHECKING that way, and I rarely ever see `assert` used outside of tests.  I'll have to keep this pattern in mind.

Reply:
Interesting! I suppose I liked catching the case at runtime just in case something exceptionally weird is happening, but I can switch to using this neat check.

Reply:
Made this change, running through pre-commit now!

Reply:
Passed pre-commit :rocket: 
Pushed to the PR branch